src/sbvr-api/sbvr-utils.coffee

```diff
@@ -353,18 +356,65 @@ getHooks = do ->
 	_getHooks = (request) ->
 		if request.resourceName?
 			resourceName = resolveSynonym(request)
-		getMethodHooks(request.method, request.vocabulary, resourceName)
+		instantiateHooks(
+			getMethodHooks(
+				request.method
+				request.vocabulary
+				resourceName
+			)
+		)
 	_getHooks.clear = -> getMethodHooks.clear()
 	return _getHooks
 
-runHook = (hookName, args) ->
+class Hook
+	constructor: (opts) ->
+		@hookFn = opts.HOOK
+		@rollbackFn = opts.ROLLBACK
+		@rollbackArgs = []
+		@executed = false
+
+	setArgument: (arg) ->
+		if @rollbackArgs.length < @rollbackFn.length
+			@rollbackArgs.push(arg)
+		else
+			throw new Error('Too many arguments supplied to rollback function')
+
+	run: (args...) ->
+		Promise.try =>
+			@hookFn(args...)
+		.tap =>
+			if @rollbackArgs.length != @rollbackFn.length
+				throw new Error('registerRollback not supplied enough arguments')
+			@executed = true
+		# If we detect that some rollbacks are registered but the rollback function is not saturated, we add the missing number of arguments as null.
+		# The ROLLBACK function should take care of checking and performing the correct number of rollback actions.
+		.tapCatch =>
+			if not _.isEmpty(@rollbackArgs)
+				missingArgs = @rollbackFn.length - @rollbackArgs.length
+				for i in _.range(missingArgs)
+					@rollbackArgs.push(null)
```

Comment:
It would be better not to fill in the extra args as null, because as null it will still throw errors if you try doing something like `null.fn()` or `null.prop`, but it will also bypass es6 default arguments, eg check out:
```js
foo = (a = { b: 'c' }) =>
	console.log(a)
foo(null) // null
foo() // {b: "c"}
foo(undefined) // {b: "c"}
```
where supplying `null` instead of the expected `undefined` will cause different behaviour (coffeescript 1 defaults apply for both null/undefined, coffeescript 2 follows the es6 defaults)

Reply:
Maybe this is not that bad after all, default arguments break the `rollback` function already as they mess with the arity count. This would help in preventing default arguments from being used there as they will essentially never work.